AX: AXPropertyValueVariant can be made smaller by avoiding std::optional

minorninth opened the pull request:
#### 9142664e45185e3ead35f0cca4477c3f23736d20
<pre>
AX: AXPropertyValueVariant can be made smaller by avoiding std::optional
<a href="https://bugs.webkit.org/show_bug.cgi?id=292802">https://bugs.webkit.org/show_bug.cgi?id=292802</a>
<a href="https://rdar.apple.com/151036216">rdar://151036216</a>

Reviewed by Tyler Wilcock.

AXPropertyValueVariant is a variant of several dozen different types,
used to store the property map for the accessibility isolated tree. If
you check sizeof(T) for each T in the variant, the largest is 16,
whereas sizeof(AXPropertyValueVariant) is 32.

It turns out that having std::optional&lt;T&gt; inside of a variant messes
with the alignment, resulting in significantly more padding. If we
stop using std::optional we can reduce the size of the variant to 24
bytes.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setOptionalProperty):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/294913@main">https://commits.webkit.org/294913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/905f36f90cb6504906238f741e16e3e9372b7d46

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22993 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13314 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108490 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53960 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105356 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23411 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31544 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78518 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35451 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106323 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93203 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58851 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17977 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11244 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53316 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87777 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11306 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110865 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30455 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22489 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87515 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89400 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87152 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22222 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32013 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9727 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24789 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35703 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30190 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33517 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31752 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco reviewed (approved):
Awesome! Should make a big difference.

I'm fine with this PR as-is. But another option to consider is using `WTF::Markable` instead of `std::optional`. It's apparently more space-efficient, while still offering optional-like semantics, so maybe it wouldn't cause the variant to grow larger than necessary.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -371,7 +372,9 @@ class AXIsolatedObject final : public AXCoreObject {
-        return propertyValue<std::optional<AccessibilityOrientation>>(AXProperty::ExplicitOrientation);
+        if (indexOfProperty(AXProperty::ExplicitOrientation) == notFound)
+            return std::nullopt;
+        return propertyValue<AccessibilityOrientation>(AXProperty::ExplicitOrientation);
```
Doesn't this do two property vector scans? I wonder if we could use the result of the `indexOfProperty` call in this function to get the value if `index != notFound`. Could require a bit of code duplication though? 🤔

minorninth replied:
Good catch - I meant to replace this with optionalAttributeValue<T> like I did elsewhere but missed this one.

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -629,9 +629,12 @@ void AXIsolatedTree::updateNodeProperties(AccessibilityObject& axObject, const A
+        case AXProperty::AXColumnIndex: {
+            std::optional<unsigned> columnIndex = axObject.axColumnIndex();
+            if (columnIndex)
+                properties.append({ AXProperty::AXColumnIndex, *columnIndex });
```
I'm fine to leave as-is, but another way to write this would be:

```cpp
if (std::optional<unsigned> columnIndex = axObject.axColumnIndex())
    properties.append({ AXProperty::AXColumnIndex, *columnIndex });
```

Same comment for the other two changes in this file.

minorninth replied:
Done, thanks!

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
```diff
@@ -369,10 +370,7 @@ class AXIsolatedObject final : public AXCoreObject {
-    {
-        return propertyValue<std::optional<AccessibilityOrientation>>(AXProperty::ExplicitOrientation);
-    }
+    virtual std::optional<AccessibilityOrientation> explicitOrientation() const { return optionalAttributeValue<AccessibilityOrientation>(AXProperty::ExplicitOrientation); }
```
Something we can address in a later PR since your PR didn't introduce this — but this function doesn't need to be virtual (nothing derives from `AXIsolatedObject`). This was my mistake, leftover from when I changed `orientation()` to `explicitOrientation()`.

minorninth committed:
AX: AXPropertyValueVariant can be made smaller by avoiding std::optional
https://bugs.webkit.org/show_bug.cgi?id=292802
rdar://151036216

Reviewed by Tyler Wilcock.

AXPropertyValueVariant is a variant of several dozen different types,
used to store the property map for the accessibility isolated tree. If
you check sizeof(T) for each T in the variant, the largest is 16,
whereas sizeof(AXPropertyValueVariant) is 32.

It turns out that having std::optional<T> inside of a variant messes
with the alignment, resulting in significantly more padding. If we
stop using std::optional we can reduce the size of the variant to 24
bytes.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::initializeProperties):
(WebCore::AXIsolatedObject::setOptionalProperty):
(WebCore::AXIsolatedObject::setProperty):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::updateNodeProperties):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/294913@main
